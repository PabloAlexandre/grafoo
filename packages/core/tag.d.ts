declare module "@grafoo/core/tag" {
  import { GrafooObject } from "@grafoo/types";

  export default function graphql(strs: TemplateStringsArray): GrafooObject;
}
