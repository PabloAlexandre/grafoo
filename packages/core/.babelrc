{
  "plugins": [
    "transform-es2015-modules-commonjs",
    ["@grafoo/loader", { "schema": "schema.graphql", "fields": ["id"] }]
  ]
}
