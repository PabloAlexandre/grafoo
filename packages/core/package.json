{
  "name": "@grafoo/core",
  "version": "0.0.1-alpha.5",
  "description": "the grafoo client core",
  "repository": "https://github.com/grafoojs/grafoo/tree/master/packages/core",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "author": "malbernaz<albernazmiguel@gmail.com>",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prepare": "grafoo-bundle --input src/index.ts",
    "test": "jest",
    "test:coverage": "jest --coverage && codecov"
  },
  "jest": {
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "json"
    ],
    "testMatch": [
      "**/__tests__/*.+(ts|tsx|js)"
    ],
    "transform": {
      "^.+\\.(ts|tsx|js)$": "<rootDir>/../../scripts/jest-setup.js"
    }
  },
  "dependencies": {
    "@grafoo/transport": "^0.0.1-alpha.5",
    "@grafoo/types": "^0.0.1-alpha.5"
  },
  "devDependencies": {
    "grafoo-bundle": "^0.0.1-alpha.5"
  },
  "gitHead": "12dae1eaac67907928d2afb723dafca1703832e5"
}
