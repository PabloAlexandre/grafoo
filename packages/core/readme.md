# `@grafoo/core`

<p><i>Grafoo core</i></p>

<p>
  <a href=https://circleci.com/gh/grafoojs/grafoo>
    <img
      src=https://img.shields.io/circleci/project/github/grafoojs/grafoo/master.svg?label=build
      alt=build
    />
  </a>
  <a href=https://codecov.io/github/grafoojs/grafoo>
    <img
      src=https://img.shields.io/codecov/c/github/grafoojs/grafoo/master.svg
      alt="coverage"
    />
  </a>
  <a href=https://github.com/grafoojs/grafoo>
    <img
      src=https://img.shields.io/npm/v/@grafoo/core.svg
      alt=npm
    >
  </a>
  <a href=https://github.com/grafoojs/grafoo>
    <img
      src=https://img.shields.io/npm/dm/@grafoo/core.svg
      alt=downloads
    >
  </a>
  <a href=https://prettier.io>
    <img
      src=https://img.shields.io/badge/code_style-prettier-ff69b4.svg
      alt="code style: prettier"
    />
  </a>
  <a href=https://lernajs.io>
    <img
      src=https://img.shields.io/badge/maintained%20with-lerna-cc00ff.svg
      alt="mantained with: lerna"
    />
  </a>
</p>

## Install

```
$ npm i @grafoo/core
```

## API

_Coming soon..._
