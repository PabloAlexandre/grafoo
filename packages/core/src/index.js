import createTransport from "@grafoo/transport/src";
import createCache from "@grafoo/cache/src";
import { assign } from "@grafoo/util";

export type * from "@grafoo/transport";
export type * from "@grafoo/cache";

export default function createClient(uri, options) {
  return assign(createCache(options), {
    request: createTransport(uri, options.fetchOptions)
  });
}
