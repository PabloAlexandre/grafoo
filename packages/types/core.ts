import { CacheOptions, CacheInstance } from "./cache";
import { TransportRequest } from "./transport";

export interface ClientInstance extends CacheInstance {
  request: TransportRequest;
}

export interface ClientOptions extends CacheOptions {
  headers: Headers;
}
