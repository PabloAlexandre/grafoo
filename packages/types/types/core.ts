import { Variables } from "./transport";

export interface ObjectsMap {
  [key: string]: { [key: string]: any };
}

export interface PathsMap {
  [key: string]: { [key: string]: any };
}

export type Listener = (objects: ObjectsMap) => void;

export interface InitialState {
  objectsMap: ObjectsMap;
  pathsMap: PathsMap;
}

export interface ClientInstance {
  request<T>(grafooObject: GrafooObject, variables?: Variables): Promise<T>;
  listen(listener: Listener): () => void;
  write: {
    (grafooObject: GrafooObject, variables: Variables, data: {}): void;
    (grafooObject: GrafooObject, data: {}): void;
  };
  read<T>(grafooObject: GrafooObject, variables?: Variables): { data?: T; objects?: ObjectsMap };
  flush(): InitialState;
}

export interface ClientOptions {
  initialState?: InitialState;
  idFields?: Array<string>;
  headers?: Headers;
}

export interface GrafooObject {
  frags?: {
    [key: string]: string;
  };
  paths: {
    [key: string]: {
      name: string;
      args: string[];
    };
  };
  query: string;
}
