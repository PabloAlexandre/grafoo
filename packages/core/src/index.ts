import createCache from "@grafoo/cache";
import createTransport from "@grafoo/transport";
import { ClientInstance, ClientOptions } from "@grafoo/types";
import { assign } from "@grafoo/util";

export default function createClient(uri: string, options?: ClientOptions): ClientInstance {
  return assign({}, createCache(options), {
    request: createTransport(uri, options && options.headers)
  });
}
