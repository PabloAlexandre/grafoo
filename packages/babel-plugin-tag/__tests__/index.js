import * as babel from "@babel/core";
import plugin from "../src";

export const transform = (program, opts) =>
  babel.transform(program, {
    plugins: [[plugin, Object.assign({ schema: "__tests__/schema.graphql" }, opts)]]
  });

afterEach(() => {
  process.env.NODE_ENV = "test";
});

describe("@grafoo/babel-plugin-tag", () => {
  it("should throw if a import is not default", () => {
    const program = 'import { gql } from "@grafoo/tag";';

    expect(() => transform(program)).toThrow();
  });

  it("should remove the imported path", () => {
    const program = 'import gql from "@grafoo/tag";';

    expect(transform(program).code).toBe("");
  });

  it("should throw if a schema is not specified", () => {
    const program = `
      import gql from "@grafoo/tag";
      const query = gql\`{ hello }\`;
    `;

    expect(() => transform(program, { schema: undefined })).toThrow();
  });

  it("should throw if a tagged template string literal has expressions in it", () => {
    const program = `
      import gql from "@grafoo/tag";
      const id = 1;
      const query = gql\`{ user(id: "\${id}") { name } }\`;
    `;

    expect(() => transform(program)).toThrow();
  });

  it("should replace a tagged template literal with the compiled grafoo object", () => {
    const program = `
      import gql from "@grafoo/tag";
      const query = gql\`
        query($start: Int!, $offset: Int!, $id: ID!) {
          posts(start: $start, offset: $offset) {
            title
            body
            createdAt
            tags { name }
            authors { name username }
          }
          user(id: $id) { name username }
        }
      \`;
    `;

    expect(transform(program, { fieldsToInsert: ["id"] }).code).toMatchSnapshot();
  });
});
