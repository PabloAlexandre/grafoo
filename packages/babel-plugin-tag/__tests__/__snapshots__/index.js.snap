// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@grafoo/babel-plugin-tag should replace a tagged template literal with the compiled grafoo object 1`] = `
"const query = {
  \\"query\\": \\"query ($id: ID!, $offset: Int!, $start: Int!) {\\\\n  posts(offset: $offset, start: $start) {\\\\n    authors {\\\\n      id\\\\n      name\\\\n      username\\\\n    }\\\\n    body\\\\n    createdAt\\\\n    id\\\\n    tags {\\\\n      id\\\\n      name\\\\n    }\\\\n    title\\\\n  }\\\\n  user(id: $id) {\\\\n    id\\\\n    name\\\\n    username\\\\n  }\\\\n}\\",
  \\"paths\\": {
    \\"posts(offset:$offset,start:$start){authors{id name username}body createdAt id tags{id name}title}\\": {
      \\"name\\": \\"posts\\",
      \\"args\\": [\\"offset\\", \\"start\\"]
    },
    \\"user(id:$id){id name username}\\": {
      \\"name\\": \\"user\\",
      \\"args\\": [\\"id\\"]
    }
  }
};"
`;
